charms/tensorboards-web-app/src/charm.py

```diff
@@ -1,158 +1,252 @@
-        super().__init__()
-
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
 
-
-class Operator(CharmBase):
+class TensorboardsWebApp(CharmBase):
     def __init__(self, *args):
+        """Initialize charm and setup the container."""
         super().__init__(*args)
 
-        self.log = logging.getLogger(__name__)
-        self.image = OCIImageResource(self, "oci-image")
-
-        for event in [
-            self.on.install,
-            self.on.upgrade_charm,
-            self.on.config_changed,
-            self.on.leader_elected,
-            self.on["ingress"].relation_changed,
-        ]:
-            self.framework.observe(event, self.main)
+        self.logger = logging.getLogger(__name__)
+        self._namespace = self.model.name
+        self._name = self.app.name
+        self._http_port = self.model.config["port"]
 
-    def main(self, event):
-        try:
+        # Explicitly use defined string since the charm can be deployed with different app name
+        self._container_name = "tensorboards-web-app"
+        self._container = self.unit.get_container(self._container_name)
+        self._k8s_resource_handler = None
 
-            self._check_leader()
-
-            interfaces = self._get_interfaces()
-
-            image_details = self._check_image_details()
+        http_port = ServicePort(int(self._http_port), name="http")
+        self.service_patcher = KubernetesServicePatch(
+            self, [http_port], service_name=f"{self._name}"
+        )
 
-        except CheckFailed as check_failed:
-            self.model.unit.status = check_failed.status
-            return
+        # Set up event handlers
+        self.framework.observe(self.on.upgrade_charm, self._on_event)
+        self.framework.observe(self.on.config_changed, self._on_event)
+        self.framework.observe(self.on.leader_elected, self._on_event)
+        self.framework.observe(self.on["ingress"].relation_changed, self._on_event)
+        self.framework.observe(self.on.tensorboards_web_app_pebble_ready, self._on_event)
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.remove, self._on_remove)
+
+    @property
+    def container(self) -> Container:
+        """Return tensorboard-controller container object."""
+        return self._container
+
+    @property
+    def k8s_resource_handler(self) -> KubernetesResourceHandler:
+        """Get the K8S resource handler."""
+        context = {
+            "app_name": self._name,
+            "namespace": self._namespace,
+            "service": self._name,
+        }
+
+        if not self._k8s_resource_handler:
+            self._k8s_resource_handler = KubernetesResourceHandler(
+                field_manager=LIGHTKUBE_FIELD_MANAGER,
+                template_files=K8S_RESOURCE_FILES,
+                context=context,
+                logger=self.logger,
+            )
```

Comment:
Take a look at the Tensorboard Controller approach towards creating the resource handler. Note how we define the `resource_types` this handler is responsible for, as well as the `labels` for identifying the managed resources: https://github.com/canonical/kubeflow-tensorboards-operator/blob/fba41aae289e978c99be6a1f02cb01958f66a6e7/charms/tensorboard-controller/src/charm.py#L151-L164

Reply:
Also note that it's convenient to hold the information required by a specific handler in a single place: https://github.com/canonical/kubeflow-tensorboards-operator/blob/fba41aae289e978c99be6a1f02cb01958f66a6e7/charms/tensorboard-controller/src/charm.py#L28-L32

Reply:
Thank you Phoevos, this is indeed a more fine-grained approach.